### PR TITLE
Internal: Unify build process to include composer install command [ED-15416]

### PR DESCRIPTION
**Pull request by Batsirai-muchareva**
## PR Checklist

- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

### CODE REVIEW

**Review by yotamselementor — DISMISSED** [review]
Are we absolutely sure that we get the same result from building the plugin with the new code as we did before the change?

File: .github/workflows/plugin-build/action.yml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+inputs:
+  REPOSITORY:
+    required: true
+    description: 'The name of the repository to build the plugin from.'
+    default: "elementor"
+  NODE_VERSION:
+    required: true
+    description: 'The version of Node.js to use.'
+    default: "20.x"
```
[the remark this answers is not in the document]
Comment (Batsirai-muchareva) [reply .github/workflows/plugin-build/action.yml:12]: When I use it in pro the action needs to generate the final artifact with the elementor-pro..etc file name

File: .github/workflows/plugin-build/action.yml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        # PATTERNS are:
+        # Everything: **/*
+        # Everything in directories starting with a period: .*/**/*
+        # Not readme.txt: !readme.txt
+        # Not changelog.txt: !changelog.txt
+        PATTERNS: |
+          **/*
+          .*/**/*
+          !readme.txt
+          !changelog.txt
```
[the remark this answers is not in the document]
Comment (Batsirai-muchareva) [reply .github/workflows/plugin-build/action.yml:44]: Do we need this @yotamselementor

File: .github/workflows/plugin-build/action.yml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        PACKAGE_VERSION=$(node -p "require('./package.json').version")
+        DATE_VERSION=$(date '+%Y%m%d.%H%M')
+        PLUGIN_FILENAME=${{ inputs.REPOSITORY }}-${PACKAGE_VERSION}-${DATE_VERSION}
+        PLUGIN_ZIP_FILENAME=${PLUGIN_FILENAME}.zip
+
+        echo "PLUGIN_ZIP_FILENAME=${PLUGIN_ZIP_FILENAME}" >> $GITHUB_ENV
+
+        npx grunt build
+
+        sed -i -E "s/Version: (.*?)/Version: ${PLUGIN_VERSION}/g" build/${{ inputs.REPOSITORY }}.php
```
[the remark this answers is not in the document]
Comment (Batsirai-muchareva) [reply .github/workflows/plugin-build/action.yml:69]: typo

**Comment by Batsirai-muchareva** [comment]
> Are we absolutely sure that we get the same result from building the plugin with the new code as we did before the change?

@yotamselementor I am completely confident that we building the same I studied all the workflow and understand how they are operating and I also tested the generated build it works as expected

**Comment by Batsirai-muchareva** [comment]
Pro is waiting for this merge so I reference the composite action here like elementor/elementor/.github/workflows/plugin-build@main targeting the main branch

File: .github/workflows/plugin-upgrade-test.yml [hunk 4]
```diff
@@ -16,7 +16,23 @@ concurrency:
...
+      - name: Checkout Source Code
+        uses: actions/checkout@v4
+
+      - name: Build Plugin
+        id: build_plugin
+        uses: ./.github/workflows/plugin-build
+
+      - name: Check if this is only a changelog PR
+        id: changelog_diffs
+        uses: ./.github/workflows/check-changelog-pr
```
Comment (davseve) [inline .github/workflows/plugin-upgrade-test.yml:34]: 1. I see that we are not using the `check-changelog-pr` output in this file
2. A small suggestion, maybe we should first check if this is a change-log PR and then run the build process?
If this is just a change-log PR there is no need to build the plugin, right?
This could save us some time.
WDYT?

File: .github/workflows/playwright.yml [hunk 5]
```diff
@@ -39,7 +39,23 @@ concurrency:
...
+      artifact_name: ${{ steps.build_plugin.outputs.artifact_name }}
+      changelog_diff: ${{ steps.changelog_diffs.outputs.changelog_diff }}
+    steps:
+      - name: Checkout Source Code
+        uses: actions/checkout@v4
+
+      - name: Build Plugin
+        id: build_plugin
+        uses: ./.github/workflows/plugin-build
+
```
Comment (davseve) [inline .github/workflows/playwright.yml:53]: Remove redundant empty line

File: action.yml [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    description: 'Optional. The version of the package to use.'
+    default: ""
+  DEPLOY_BUILDS_ENDPOINT:
+    required: true
+    description: 'The endpoint to deploy the builds to.'
+    default: ""
+  REPOSITORY_NAME:
+    required: true
+    description: 'The name of the repository.'
+    default: ""
```
Comment (davseve) [inline action.yml:16]: I see that you have a few required input.vars
In the action call I see that you don't always pass them.
Maybe it is better to make them optional.
E.g, https://github.com/elementor/elementor/pull/28308/files#diff-1b5aac1ec85d5af43f331dd9f3142bdb3481d67809cb8d2bbeca281e40552fd9R35

File: action.yml
Comment (davseve) [inline action.yml:1]: Can we place this file inside the .github dir?
Only if possible.

File: .github/workflows/latest-release.yml [hunk 7]
```diff
@@ -60,13 +61,72 @@ jobs:
...
+      - name: Update Beta tag in Readme.txt file
+        run: |
+          if grep -q "Beta tag:" ./readme.txt
+            then
+              echo "Replace existing beta tag in readme file"
+              sed -i -E "s/Beta tag: .*/Beta tag: ${{ env.BETA_VERSION }}/g" ./readme.txt
+            else
+              echo "Add beta tag to readme file"
+              sed -i -E '/^Stable tag: .*/a\'$'\n'"Beta tag: ${{ env.BETA_VERSION }}$(printf '\n\r')" ./readme.txt
+          fi
```
Comment (davseve) [inline .github/workflows/latest-release.yml:117]: Not sure that this should be here
There is a comment regarding this in the other location.

**Review by davseve — COMMENTED** [review]
I added a few comments
Lets talk about it tommorow.

File: .github/workflows/plugin-build.yml [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        run: |
+          PLUGIN_FILENAME=elementor-${{ env.BUILD_PACKAGE_VERSION }}
+          PLUGIN_ZIP_FILENAME=${PLUGIN_FILENAME}.zip
+
+          mv build elementor
+          zip -r ${PLUGIN_ZIP_FILENAME} elementor
+
+          echo "PLUGIN_FILENAME=${PLUGIN_FILENAME}" >> $GITHUB_ENV
+
+
```
Comment (davseve) [inline .github/workflows/plugin-build.yml:164]: Remove redundant empty line

File: .github/workflows/one-click-release.yml [hunk 9]
```diff
@@ -25,63 +25,112 @@ permissions:
...
+          POSTFIX: ''
+          OVERRIDE_PACKAGE_VERSION: true
+
+  build-plugin:
+    name: Build Plugin
+    needs: [configure-env-variables]
+    uses: ./.github/workflows/plugin-build.yml
+    with:
+      PACKAGE_VERSION: ${{ needs.configure-env-variables.outputs.PACKAGE_VERSION }}
+      PACKAGE_FILES_VERSION: ${{ needs.configure-env-variables.outputs.PACKAGE_FILES_VERSION }}
```
Comment (davseve) [inline .github/workflows/one-click-release.yml:66]: WDYT?
```suggestion
      PACKAGE_FILES_VERSION: ${{ needs.configure-env-variables.outputs.PACKAGE_FILES_VERSION }}
      IS_RELEASE: true
```

File: .github/workflows/playwright.yml [hunk 10]
```diff
@@ -37,14 +37,30 @@ concurrency:
...
+      - name: Check if this is only a changelog PR
+        id: changelog_diffs
+        uses: ./.github/workflows/check-changelog-pr
+
   build-plugin:
-    name: Build plugin
-    uses: ./.github/workflows/build.yml
+    name: Build Plugin
+    needs: [validate-changelog-diff]
+    if: ${{ github.event.pull_request.title == null || needs.validate-changelog-diff.outputs.changelog_diff }}
```
[the remark this answers is not in the document]
Comment (Batsirai-muchareva) [reply .github/workflows/playwright.yml:56]: @davseve I am not sure its a condition which was there previously can we remove it WDYT?

File: .github/workflows/plugin-build.yml [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+      IS_RELEASE:
+        required: false
+        type: boolean
+        description: 'Optional. If true, the plugin build will be released.'
+        default: false
+      IS_CORE_FOR_PRO:
+        required: false
+        type: boolean
+        description: 'Optional. If true, the core for pro plugin will be built.'
+        default: false
```
[the remark this answers is not in the document]
Comment (Batsirai-muchareva) [reply .github/workflows/plugin-build.yml:36]: @davseve because we have 2 execution of the build for pro and build for core IS_CORE_FOR_PRO is not the same with custom core

File: .github/workflows/plugin-build.yml [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+
+          npx grunt build
+
+      - name: Zip Elementor Core Build Files
+        if: ${{ env.REPOSITORY_NAME == 'elementor' }}
+        run: |
+          PLUGIN_FILENAME=elementor-${{ env.BUILD_PACKAGE_VERSION }}
+          PLUGIN_ZIP_FILENAME=${PLUGIN_FILENAME}.zip
+
+          mv build elementor
```
[the remark this answers is not in the document]
Comment (Batsirai-muchareva) [reply .github/workflows/plugin-build.yml:165]: it's just renaming the folder so that when you unzip it its has the name elementor not build

File: .github/workflows/one-click-release.yml (hunk 9 above)
Comment (Batsirai-muchareva) [reply .github/workflows/one-click-release.yml:66]: Yes I had missed it

File: .github/workflows/latest-release.yml [hunk 13]
```diff
@@ -60,13 +60,64 @@ jobs:
...
           OVERRIDE_PACKAGE_VERSION: true
+
+  build-plugin:
+    if: github.event.pull_request.merged == true && github.base_ref == 'main' && !contains(github.event.pull_request.labels.*.name, 'skip release' )
+    name: Build Plugin
+    needs: [configure-env-variables]
+    uses: ./.github/workflows/plugin-build.yml
+    with:
+      PACKAGE_VERSION: ${{ needs.configure-env-variables.outputs.PACKAGE_VERSION }}
+      PACKAGE_FILES_VERSION: ${{ needs.configure-env-variables.outputs.PACKAGE_FILES_VERSION }}
```
[the remark this answers is not in the document]
Comment (Batsirai-muchareva) [reply .github/workflows/latest-release.yml:71]: Yes I had missed it

**Comment by Batsirai-muchareva** [comment]
After this review I am going to test the builds generated

File: .github/workflows/playwright.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (davseve) [reply .github/workflows/playwright.yml:56]: this just looks wired
If there is not title or it has change-log than run the build process?

File: .github/workflows/playwright.yml [hunk 14]
```diff
@@ -37,14 +37,30 @@ concurrency:
...
+    uses: ./.github/workflows/plugin-build.yml
+
   Playwright:
     name: Playwright test - ${{ matrix.shardIndex }} on PHP 8.0
     runs-on: ubuntu-latest
-    needs: [build-plugin]
-    if: ${{ github.event.pull_request.title == null || needs.build-plugin.outputs.changelog_diff }}
+    needs: [build-plugin, validate-changelog-diff]
+    if: ${{ github.event.pull_request.title == null || needs.validate-changelog-diff.outputs.changelog_diff }}
```
[the remark this answers is not in the document]
Comment (davseve) [reply .github/workflows/playwright.yml:63]: See the comment above

File: .github/workflows/plugin-build.yml [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+        description: 'Optional. The version of the package to use.'
+        default: ""
+      PACKAGE_FILES_VERSION:
+        required: false
+        type: string
+        default: ""
+      ELEMENTOR_PRO_CORE_BRANCH:
+        required: false
+        type: string
+        description: 'The branch of the core repository to use.'
```
[the remark this answers is not in the document]
Comment (davseve) [reply .github/workflows/plugin-build.yml:20]: I think this is not relevant anymore. 
We have custom core that replaces it.

[the remark this answers is not in the document]
Comment (Batsirai-muchareva) [reply .github/workflows/plugin-build.yml:20]: @davseve `ELEMENTOR_PRO_CORE_BRANCH` is important because in custom core we get a branch from input of manual dispatch so we want to be able to pass that branch to the build process

File: .github/workflows/plugin-build.yml [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+          cd $ROOT_DIR
+
+          unzip -qo ${PLUGIN_ZIP_PATH} -d ./elementor-pro
+
+          echo "PLUGIN_ZIP_FILENAME=${PLUGIN_ZIP_FILENAME}" >> $GITHUB_ENV
+          echo "PLUGIN_FILENAME=${PLUGIN_FILENAME}" >> $GITHUB_ENV
+
+      - name: Publish zip on push to release or develop
+        shell: bash
+        if: env.REPOSITORY_NAME == 'elementor-pro' && github.event_name == 'push' && startsWith(github.repository, 'elementor/') && (startsWith(github.ref, 'refs/heads/release/') || github.ref == 'refs/heads/develop')
```
[the remark this answers is not in the document]
Comment (Batsirai-muchareva) [reply .github/workflows/plugin-build.yml:197]: @davseve  is this step needed let me know I am not understanding cause the endpoint is passed so i dont know where its going. ``` php       - name: Publish zip on push to release or develop
        shell: bash
        if: env.REPOSITORY_NAME == 'elementor-pro' && github.event_name == 'push' && startsWith(github.repository, 'elementor/') && (startsWith(github.ref, 'refs/heads/release/') || github.ref == 'refs/heads/develop')
        run: curl --fail -F "package=@${{ env.PLUGIN_ZIP_FILENAME }}" "${{ secrets.DEPLOY_BUILDS_ENDPOINT }}&type=pro"
 ```

File: .github/workflows/playwright.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (Batsirai-muchareva) [reply .github/workflows/playwright.yml:56]: @davseve so I think we need something like this right `if: ${{ github.event.pull_request.title != 'changelog' && !needs.validate-changelog-diff.outputs.changelog_diff }}
`

File: .github/workflows/plugin-build.yml (hunk 16 above)
[the remark this answers is not in the document]
Comment (davseve) [reply .github/workflows/plugin-build.yml:201]: I think that this step is not valid anymore
In the past all the release branches were under the release prefix E.g, 'release/3.20.0`.
As well we had a develop branch, 
Every release used to get merged to the develop branch so that it always used to include the last updated release. 
@louiswol94 WDYT?
Do we still need this step?

File: .github/workflows/plugin-upgrade-test.yml [hunk 17]
```diff
@@ -15,40 +15,65 @@ concurrency:
...
+    if: ${{ github.event.pull_request.title == null || needs.validate-changelog-diff.outputs.changelog_diff }}
+    uses: ./.github/workflows/plugin-build.yml
+
   run-upgrade-test:
     name: Playwright plugin upgrade test
     runs-on: ubuntu-latest
-    needs: [build-plugin]
-    if: ${{ github.event.pull_request.title == null || needs.build-plugin.outputs.changelog_diff }}
+    needs: [build-plugin, validate-changelog-diff]
+    if: ${{ github.event.pull_request.title == null || needs.validate-changelog-diff.outputs.changelog_diff }}
```
Comment (davseve) [inline .github/workflows/plugin-upgrade-test.yml:41]: I think that all the `changelog_diff` conditions should be the same.
Please go over all of them and unify them id necessary.

File: .github/workflows/plugin-build.yml [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+        default: ""
+      PACKAGE_FILES_VERSION:
+        required: false
+        type: string
+        default: ""
+      ELEMENTOR_PRO_CORE_BRANCH:
+        required: false
+        type: string
+        description: 'The branch of the core repository to use when executing this action from Elementor Pro.'
+        default: "main"
```
[the remark this answers is not in the document]
Comment (Batsirai-muchareva) [reply .github/workflows/plugin-build.yml:21]: Its being used here 
`
      - name: Clone Elementor Core For Pro
        if: env.REPOSITORY_NAME == 'elementor-pro'
        run: |
          ELEMENTOR_CORE_BRANCH="${{ inputs.ELEMENTOR_PRO_CORE_BRANCH }}"

          rm -rf ../elementor
          echo "Cloning ELEMENTOR_CORE_BRANCH: ${ELEMENTOR_CORE_BRANCH}"

          # Cloning the core plugin into the elementor folder outside elementor-pro folder
          git clone --single-branch --branch ${ELEMENTOR_CORE_BRANCH} https://github.com/elementor/elementor.git ../elementor
`

File: .github/workflows/plugin-build.yml [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+      - name: Prevent git tag versioning
+        if: env.REPOSITORY_NAME == 'elementor' && inputs.IS_RELEASE == false
+        run: npm config set git-tag-version false
+
+      - name: Skip puppeteer download
+        if: env.REPOSITORY_NAME == 'elementor-pro' && inputs.IS_RELEASE == false
+        run: export PUPPETEER_SKIP_DOWNLOAD=true
+
+      - name: Clone Elementor Core For Pro
+        if: env.REPOSITORY_NAME == 'elementor-pro'
```
[the remark this answers is not in the document]
Comment (Batsirai-muchareva) [reply .github/workflows/plugin-build.yml:127]: @davseve any pro npx build needs CORE else it will fail, grunt for PRO is dependent on CORE

File: .github/workflows/plugin-build.yml (hunk 16 above)
[the remark this answers is not in the document]
Comment (louiswol94) [reply .github/workflows/plugin-build.yml:201]: I don't think we need it.

[the remark this answers is not in the document]
Comment (davseve) [reply .github/workflows/plugin-build.yml]: @Batsirai-muchareva 
Than this condition can be deleted

[the remark this answers is not in the document]
Comment (Batsirai-muchareva) [reply .github/workflows/plugin-build.yml]: @davseve removed

File: .github/workflows/check-changelog-pr/action.yml
Comment (yotamselementor) [inline .github/workflows/check-changelog-pr/action.yml:1]: This seems like something that should be done in a separate PR

File: .github/workflows/build-plugin/action.yml
Comment (yotamselementor) [inline .github/workflows/build-plugin/action.yml:1]: Wouldn't the PR be much simpler and, more importantly, safer, if we just use this composite action in all actions?
This is the composite action that the `one-click-release` action uses, so if `one-click-release` doesn't change, we don't have any major risk. The rest of the actions can align to use it.
Later, if we see a need, we could continue the refactor in a separate PR.

**Review by yotamselementor — CHANGES_REQUESTED** [review]

File: .github/workflows/build-plugin/action.yml
Comment (Batsirai-muchareva) [reply .github/workflows/build-plugin/action.yml:1]: @yotamselementor We are also calling this build in the PRO plugin so that's why we go for the reusable workflow instead of the composite action

File: .github/workflows/check-changelog-pr/action.yml
Comment (Batsirai-muchareva) [reply .github/workflows/check-changelog-pr/action.yml:1]: @yotamselementor Because of the changes of having a unified build we created it in a separate PR so that the build can have one purpose and change log check another purpose in a separate file
